python/ray/serve/_private/http_proxy.py

```diff
@@ -830,11 +840,16 @@ async def send_request_to_replica_streaming(
                         curr_time_s=time.time(),
                     ),
                 )
-            except TimeoutError:
+            except TimeoutError as is_first_message:
                 logger.warning(
                     f"Request {request_id} timed out after "
                     f"{self.request_timeout_s}s while executing."
                 )
+                # We should only send timeout response if we have not sent
+                # any messages to the client yet. Header (including status code)
+                # messages can only be sent once.
+                if str(is_first_message) == str(True):
```

Comment:
what the heck is going on here? if we want to return custom metadata in the exception please make a wrapper exception type instead of this hackery